### PR TITLE
Custom analyzer to reject  unknown parameters

### LINKED ISSUE #85710: Create analyzer accepting any random param and returning that in response as well
### Elasticsearch Version

8.1.1

### Installed Plugins

None

### Java Version

_bundled_

### OS Version

macos

### Problem Description

Create a custom analyzer using below request, **Note `foo` as an param to analyzer with `bar` value.**

```
PUT customanlyzerindex
{
    "settings": {
        "analysis": {
            "analyzer": {
                "my_analyzer": {
                    "type": "custom",
                    "tokenizer": "standard",
                    "foo": "bar"
                }
            }
        }
    }
}
```

Index is created successfully, and `foo` param is returned in the _settings API response.

GET customanlyzerindex/_settings

```
{
    "customanz": {
        "settings": {
            "index": {
                "routing": {
                    "allocation": {
                        "include": {
                            "_tier_preference": "data_content"
                        }
                    }
                },
                "number_of_shards": "1",
                "provided_name": "customanz",
                "creation_date": "1649229792216",
                "analysis": {
                    "analyzer": {
                        "my_analyzer": {
                            "type": "custom",
                            "foo": "bar",
                            "tokenizer": "standard"
                        }
                    }
                },
                "number_of_replicas": "1",
                "uuid": "zo-99vRVQD6igTzbyFMNKA",
                "version": {
                    "created": "8010199"
                }
            }
        }
    }
}
```

### Steps to Reproduce

Same as problem Description 

### Logs (if relevant)

N/A

### PR DESCRIPTION
**Pull request by likzn**
# Main Change
Nowtime, when we create custom analyzer with some not required params("foo":"nar"), it will succeed. This PR will make check with these parms to ensure these were our need.

Close: #85710

```
PUT customanlyzerindex
{
    "settings": {
        "analysis": {
            "analyzer": {
                "my_analyzer": {
                    "type": "custom",
                    "tokenizer": "standard",
                    "foo": "bar"
                }
            }
        }
    }
}
```

### CODE REVIEW

**Comment by likzn** [comment]
@elasticsearchmachine test this please

**Comment by likzn** [comment]
@javanna @romseygeek @jtibshirani Hi, PTAL  if free~

**Comment by likzn** [comment]
Hi, can someone take a look about it

**Comment by likzn** [comment]
@javanna Hi, PTAL

File: server/src/main/java/org/elasticsearch/index/analysis/AnalyzerComponents.java [hunk 1]
```diff
@@ -44,6 +44,12 @@ static AnalyzerComponents createComponents(
         final Map<String, CharFilterFactory> charFilters,
         final Map<String, TokenFilterFactory> tokenFilters
     ) {
+        for (String key : analyzerSettings.keySet()) {
+            switch (key) {
+                case "tokenizer", "char_filter", "filter", "type", "position_increment_gap" -> {}
+                default -> throw new IllegalArgumentException("Custom Analyzer not support [" + key + "] now");
```
Comment (javanna) [inline server/src/main/java/org/elasticsearch/index/analysis/AnalyzerComponents.java:50]: could you rephrase the error to `"Custom analyzer [" + name + "] does not support [" + key + "]"` ?

File: modules/analysis-common/src/yamlRestTest/resources/rest-api-spec/test/search.query/10_match.yml [hunk 2]
```diff
@@ -16,7 +16,6 @@
                   tokenizer: standard
                   filter: [lowercase]
                 search:
-                  rest_total_hits_as_int: true
```
Comment (javanna) [inline modules/analysis-common/src/yamlRestTest/resources/rest-api-spec/test/search.query/10_match.yml:19]: why was this change necessary?

File: modules/analysis-common/src/yamlRestTest/resources/rest-api-spec/test/indices.analyze/10_analyze.yml [hunk 3]
```diff
@@ -59,3 +59,22 @@
...
+            analysis:
+              analyzer:
+                my_analyzer:
+                  type: custom
+                  tokenizer: standard
+                  foo: bar
+
+  - match: { status: 400 }
+  - match: { error.type: illegal_argument_exception }
+  - match: { error.reason: "Custom Analyzer not support [foo] now" }
```
Comment (javanna) [inline modules/analysis-common/src/yamlRestTest/resources/rest-api-spec/test/indices.analyze/10_analyze.yml:80]: ++ thanks for adding this test!

File: server/src/main/java/org/elasticsearch/index/analysis/AnalyzerComponents.java [hunk 4]
```diff
@@ -44,6 +44,12 @@ static AnalyzerComponents createComponents(
         final Map<String, CharFilterFactory> charFilters,
         final Map<String, TokenFilterFactory> tokenFilters
     ) {
+        for (String key : analyzerSettings.keySet()) {
+            switch (key) {
+                case "tokenizer", "char_filter", "filter", "type", "position_increment_gap" -> {}
```
Comment (javanna) [inline server/src/main/java/org/elasticsearch/index/analysis/AnalyzerComponents.java:49]: another way of doing this, without deduplicating the expected analyzer keys, would be to duplicate the settings into a mutable map, which we'd remove from every time we read a certain setting. At the end, if any item is left in the map it means that some unsupported param has been provided.

File: server/src/test/java/org/elasticsearch/index/analysis/AnalysisTests.java [hunk 5]
```diff
@@ -103,4 +116,15 @@ public void testParseWordList() throws IOException {
...
         assertEquals(Arrays.asList("hello", "world"), wordList);
     }
+
+    public void testCustomAnalyzerWithNotSupportKey() {
+        Settings analyzerSettings = Settings.builder().put("tokenizer", "standard").put("foo", "bar").put("type", "custom").build();
+
+        try {
+            createComponents("my_analyzer", analyzerSettings, testAnalysis.tokenizer, testAnalysis.charFilter, testAnalysis.tokenFilter);
+            fail("expected failure");
+        } catch (IllegalArgumentException e) {
```
Comment (javanna) [inline server/src/test/java/org/elasticsearch/index/analysis/AnalysisTests.java:126]: you can use expectThrows instead .

File: server/src/test/java/org/elasticsearch/index/analysis/AnalysisTests.java [hunk 6]
```diff
@@ -27,9 +31,18 @@
...
 import static org.hamcrest.Matchers.is;
 
 public class AnalysisTests extends ESTestCase {
+    private static TestAnalysis testAnalysis;
+    private static Settings settings = Settings.builder().put(IndexMetadata.SETTING_VERSION_CREATED, Version.CURRENT).build();
+
+    @BeforeClass
+    public static void setup() throws IOException {
+        testAnalysis = createTestAnalysis(new Index("test", "_na_"), settings);
+    }
```
Comment (javanna) [inline server/src/test/java/org/elasticsearch/index/analysis/AnalysisTests.java:44]: do you need this as before class? It looks like it's only used in the new method that you introduced?

File: server/src/test/java/org/elasticsearch/index/analysis/AnalysisTests.java [hunk 7]
```diff
@@ -103,4 +116,15 @@ public void testParseWordList() throws IOException {
         List<String> wordList = Analysis.getWordList(env, nodeSettings, "foo.bar");
         assertEquals(Arrays.asList("hello", "world"), wordList);
     }
+
+    public void testCustomAnalyzerWithNotSupportKey() {
```
Comment (javanna) [inline server/src/test/java/org/elasticsearch/index/analysis/AnalysisTests.java:120]: testCustomAnalyzerWithUnsupportedKey ?

**Review by javanna — COMMENTED** [review]
I left a couple of comments, the approach looks good to me. I am going to run tests and edit the title which is going to go in the changelog entry.

**Comment by javanna** [comment]
@elasticsearchmachine test this please

File: docs/changelog/89112.yaml [hunk 8]
```diff
@@ -0,0 +1,6 @@
+pr: 89112
+summary: Make param check when create custom analyzer
+area: Search/Analysis
+type: enhancement
+issues:
+ - 85710
```
Comment (javanna) [inline docs/changelog/89112.yaml:6]: the changelog generation is automatic in most situations, no need to create the entry yourself. And it gets updated as the labels and title of the PR get updated. Could you remove the changelog from your PR, then it should get automatically created with the expected area and all other fields. The current build errors are around the wrong area.

File: server/src/main/java/org/elasticsearch/index/analysis/AnalyzerComponents.java [hunk 9]
```diff
@@ -44,6 +44,12 @@ static AnalyzerComponents createComponents(
         final Map<String, CharFilterFactory> charFilters,
         final Map<String, TokenFilterFactory> tokenFilters
     ) {
+        for (String key : analyzerSettings.keySet()) {
```
Comment (javanna) [inline server/src/main/java/org/elasticsearch/index/analysis/AnalyzerComponents.java:47]: one additional suggestion: this is technically a breaking change, as existing indices will not be loaded if they have have an unknown param in the definition of a custom analyzer. We could mitigate this by making the new behaviour depend on the index created version, so that we'd throw error only for newly created indices. We'd need to have the index created version propagated from AnalysisRegistry though.

Comment (likzn) [reply server/src/main/java/org/elasticsearch/index/analysis/AnalyzerComponents.java:47]: Sorry. I am late. But i find the `method of createComponents` is stateless. How can we propagate the index created version to it?

Comment (likzn) [reply server/src/main/java/org/elasticsearch/index/analysis/AnalyzerComponents.java:47]: OK. I got it.

**Comment by likzn** [comment]
@javanna Hi, i resolved all comments. PTAL~

**Comment by javanna** [comment]
@elasticmachine update branch

**Comment by javanna** [comment]
retest this please

**Comment by benwtrent** [comment]
@javanna I am not sure we should actually make this change.

All analyzer builders accept unknown parameters without throwing, and it has been that way since the start. Seems like an expensive and broad ranging breaking change that might not be worth the squeeze.

**Comment by javanna** [comment]
I agree that it's a breaking change. I agree it is rather late to fix it. It does feel like something we'll want to fix though at some point, in that incorrect configuration is being silently accepted, as opposed to how Elasticsearch works in many other places. Ideally we'd promptly fail instead and provide feedback to users.

**Comment by benwtrent** [comment]
@javanna the real fix is to implement something like: 

https://github.com/elastic/elasticsearch/issues/41299#issuecomment-3179687244

And provide a deprecation warning on unknown parameters and deprecate the old ctor for custom analyzers that do not provide a "known settings" value.

I am closing this PR as it is very far away from a good path forward. 

New work can be reopened through the plan laid out in the linked issue.

**Comment by javanna** [comment]
Sounds good to me @benwtrent thanks
